File: sdk/cpp/src/codec.cpp
#include "cap/codec.h"

#include <google/protobuf/io/coded_stream.h>
#include <google/protobuf/io/zero_copy_stream_impl_lite.h>

namespace cap {

std::vector<uint8_t> MarshalDeterministic(const google::protobuf::Message& msg) {
  const size_t size = msg.ByteSizeLong();
  std::vector<uint8_t> buf(size);
  if (size == 0) return buf;

  google::protobuf::io::ArrayOutputStream array_stream(buf.data(),
                                                        static_cast<int>(size));
  google::protobuf::io::CodedOutputStream coded_stream(&array_stream);
  coded_stream.SetSerializationDeterministic(true);
  if (!msg.SerializeToCodedStream(&coded_stream)) {
    return {};
  }
  return buf;
}

std::vector<uint8_t> MarshalUnsignedForSignature(
    const cordum::agent::v1::BusPacket& packet) {
  cordum::agent::v1::BusPacket clone;
  clone.CopyFrom(packet);
  clone.clear_signature();
  return MarshalDeterministic(clone);
}

}  // namespace cap

File: sdk/php/src/Codec.php
<?php

declare(strict_types=1);

namespace Cordum\Cap;

use Cordum\Agent\V1\BusPacket;
use Google\Protobuf\Internal\Message;

/**
 * Deterministic protobuf serialization utilities.
 */
final class Codec
{
    /**
     * Serialize a protobuf message deterministically.
     * Google's PHP protobuf runtime produces deterministic output for proto3 messages.
     */
    public static function marshalDeterministic(Message $msg): string
    {
        return $msg->serializeToString();
    }

    /**
     * Serialize a BusPacket with the signature field cleared.
     * Used as the input to ECDSA signing/verification.
     */
    public static function marshalUnsignedForSignature(BusPacket $packet): string
    {
        $clone = clone $packet;
        $clone->setSignature('');
        return self::marshalDeterministic($clone);
    }
}
